### PR TITLE
Fixing incorrect text reporting when hovering mouse over Scintilla controls.

### LINKED ISSUE #5450: NVDA problem with Notepad++
**Reported by DataTriny on 2015-10-30 23:11**
Hi,
When hovering the mouse on Notepad++ text editor, NVDA reports the wrong line (precisely it reports the fifth line below).

### PR DESCRIPTION
**Pull request by DataTriny**
### Link to issue number:

Fixes #5450.

### Summary of the issue:

When hovering the mouse over Notepad++ text editor, and over any Scintilla control in general, NVDA fails to report the line below the mouse cursor.

### Description of how this pull request fixes the issue:

Converts from screen coordinates to client coordinates, in `ScintillaTextInfo._getOffsetFromPoint`, so that we can get accurate text offset.

### Testing performed:

Created a text file containing several lines in Notepad++, hovered the mouse over each line and checked that the reported text was right. Tested both with Notepad++'s window in full screen, maximized as well as in a smaller size.

### Known issues with pull request:

None

### Change log entry:

Bug fixes

+ NVDA now properly reports the line below the mouse cursor in Notepad++ and other Scintilla based editors.

### CODE REVIEW

File: source/NVDAObjects/window/scintilla.py [hunk 1]
```diff
@@ -66,6 +66,7 @@ class TextRangeStruct(ctypes.Structure):
 class ScintillaTextInfo(textInfos.offsets.OffsetsTextInfo):
 
 	def _getOffsetFromPoint(self,x,y):
+		x, y = winUser.ScreenToClient(self.obj.windowHandle, x, y)
 		return watchdog.cancellableSendMessage(self.obj.windowHandle,SCI_POSITIONFROMPOINT,x,y)
```
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]: Have you considered using SCI_CHARPOSITIONFROMPOINT or even SCI_CHARPOSITIONFROMPOINTCLOSE here? The latter allows you to check whether the retrieved point is valid.

Note that it is very important to test this code with high dpi, i.e. setting the DPI to 125 or higher. Does it still function properly in that case?

File: source/NVDAObjects/window/scintilla.py [hunk 2]
```diff
@@ -66,6 +66,7 @@ class TextRangeStruct(ctypes.Structure):
 class ScintillaTextInfo(textInfos.offsets.OffsetsTextInfo):
 
 	def _getOffsetFromPoint(self,x,y):
+		x, y = winUser.ScreenToClient(self.obj.windowHandle, x, y)
 		return watchdog.cancellableSendMessage(self.obj.windowHandle,SCI_POSITIONFROMPOINT,x,y)
 
 	def _getPointFromOffset(self,offset):
```
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]: Could you please verify whether this method works as expected? Note that there is a long standing bug in this method (i.e. the `if point.x and point.y:` line). Please change this into `if point.x is not None and point.y is not None:`

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Thanks for taking this

File: source/NVDAObjects/window/scintilla.py (hunk 1 above)
Comment (DataTriny) [reply source/NVDAObjects/window/scintilla.py]: The problem with `SCI_POSITIONFROMPOINTCLOSE`, `SCI_CHARPOSITIONFROMPOINT` and `SCI_CHARPOSITIONFROMPOINTCLOSE` is that they wont return a valid position if the mouse is placed after the end of the line, thus making it inconsistent compared to how NVDA behaves (and how it should behave) on other editable controls.
Apart from that, I've tested with a DPI of 125% and it works just fine.

**Comment by dingpengyu** [comment]
hi
@DataTriny 
could you please merge master to fix the failing tests?

**Comment by DataTriny** [comment]
Thanks @dingpengyu for the tip!
@leonardder Are your concerns addressed?

**Comment by dingpengyu** [comment]
> Thanks @dingpengyu for the tip!
> @leonardder Are your concerns addressed?
Yes, the problem has been solved. In addition, when does this PR start to review? Really this PR is very useful in np++.

**Comment by DataTriny** [comment]
Actually the delay is mostly my fault as I didn't see the review comments immediately (thanks GitHub notifications).

**Comment by michaelDCurran** [comment]
@leonardder  changes are requested from you. Is that still correct?

**Review by LeonarddeR — APPROVED** [review]
Sorry, I missed the fixes to this. It's now ok, go ahead :)

**Review by michaelDCurran — APPROVED** [review]

**Comment by DrSooom** [comment]
@michaelDCurran: Please change the milestone from 2019.1 to 2019.2. Thanks.

**Pull request merged**